pandas/_libs/lib.pyx

```diff
@@ -2875,10 +2874,59 @@ def map_infer_mask(
     mask : ndarray
         uint8 dtype ndarray indicating values not to apply `f` to.
     convert : bool, default True
-        Whether to call `maybe_convert_objects` on the resulting ndarray
+        Whether to call `maybe_convert_objects` on the resulting ndarray.
     na_value : Any, optional
         The result value to use for masked values. By default, the
-        input value is used
+        input value is used.
+    dtype : numpy.dtype
+        The numpy dtype to use for the result ndarray.
+
+    Returns
+    -------
+    np.ndarray
+    """
+    # Passed so we can use infused types depending on the result dtype
+    dummy = np.empty(0, dtype=dtype)
```

Comment:
I think you can also just pass a pointer to avoid python runtime interaction

Reply:
I'm not familiar - can you spell out the syntax here?

Reply:
See if this works

http://docs.cython.org/en/latest/src/userguide/fusedtypes.html#fused-types-and-arrays

Reply:
Cannot take address of Python variable 'dummy'

Reply:
probably need to cdef it as ndarray

Reply:
Yea shouldn't even need an array. Typing from phone so sorry for grammar but should be

cdef numeric_object_t dummy

Then when calling function use address of operator &dummy

The callee should have numeric_object_t *dummy as an argument

Reply:
Ah OK. That's unfortunate - I think @jbrockmendel has the best suggestion then to make `result` caller allocated

Reply:
> I think @jbrockmendel has the best suggestion then to make `result` caller allocated

I misunderstood that suggestion before - I think this would only be doable if we are okay with also moving `convert` out of the function. So instead of something like:

```
result = lib.map_infer_mask(arr, f, mask.view(np.uint8), map_convert)
```

we are now doing

```
result = np.empty_like(arr, dtype="object")
lib.map_infer_mask(result, arr, f, mask.view(np.uint8))
if map_convert:
	result = maybe_convert_objects(result)
```

This seems worse to me (duplicating the convert code), but happy to make the change if that's what's preferred by others.

Reply:
> this would only be doable if we are okay with also moving convert out of the function

yes im fine with this.  though i think if you called the result something other than `result` you might not have the same TypeError issues, and since we're not iterating over it its no big deal perf-wise